File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/selection/handlers/KotlinDefaultSelectionHandler.java
package org.jetbrains.kotlin.ui.editors.selection.handlers;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;

public class KotlinDefaultSelectionHandler implements KotlinElementSelectionHandler {
    
    @Override
    public boolean canSelect(PsiElement enclosingElement) {
        return true;
    }
    
    @Override
    public TextRange selectEnclosing(PsiElement enclosingElement, TextRange selectedRange) {
        return enclosingElement.getTextRange();
    }
    
    @Override
    public TextRange selectPrevious(PsiElement enclosingElement, PsiElement selectionCandidate, TextRange selectedRange) {
        return new TextRange(selectionCandidate.getTextRange().getStartOffset(), selectedRange.getEndOffset());
    }
    
    @Override
    public TextRange selectNext(PsiElement enclosingElement, PsiElement selectionCandidate, TextRange selectedRange) {
        return new TextRange(selectedRange.getStartOffset(), selectionCandidate.getTextRange().getEndOffset());
    }
    
}

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/selection/handlers/KotlinDefaultSelectionHandler.kt
package org.jetbrains.kotlin.ui.editors.selection.handlers

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement

public open class KotlinDefaultSelectionHandler:KotlinElementSelectionHandler {
    
    override fun canSelect(enclosingElement:PsiElement)
        = true

    override fun selectEnclosing(enclosingElement:PsiElement, selectedRange:TextRange)
        = enclosingElement.getTextRange()

    override fun selectPrevious(enclosingElement:PsiElement, selectionCandidate:PsiElement, selectedRange:TextRange)
        = selectionWithElementAppendedToBeginning(selectedRange, selectionCandidate)

    override fun selectNext(enclosingElement:PsiElement, selectionCandidate:PsiElement, selectedRange:TextRange)
        = selectionWithElementAppendedToEnd(selectedRange, selectionCandidate)

     protected fun selectionWithElementAppendedToEnd(selection: TextRange, element: PsiElement):TextRange
     	= TextRange(selection.getStartOffset(), element.endOffset)
    
    protected fun selectionWithElementAppendedToBeginning(selection: TextRange, element: PsiElement):TextRange
     	= TextRange(element.startOffset, selection.getEndOffset())
}